osu.Game/Rulesets/Objects/Types/IHasEndTime.cs

```diff
@@ -1,24 +1,12 @@
-﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
-using Newtonsoft.Json;
+using System;
 
 namespace osu.Game.Rulesets.Objects.Types
 {
-    /// <summary>
-    /// A HitObject that ends at a different time than its start time.
-    /// </summary>
-    public interface IHasEndTime
+    [Obsolete("Use IHasDuration instead.")] // can be removed 20201126
```

Comment:
This is in the opposite direction and will break user rulesets that already use `IHasEndTime`, rendering this obsoletion   notice useless. E.g.:

https://github.com/EVAST9919/bosu/blob/33ea85fef20340ab24168aa4f4372a46fa5df34d/osu.Game.Rulesets.Bosu/Beatmaps/BosuBeatmapConverter.cs#L45

https://github.com/LumpBloom7/sentakki/blob/2a637af0127fa17d7d8b07c5b199301377d777d4/osu.Game.Rulesets.Sentakki/Beatmaps/SentakkiBeatmapConverter.cs#L47

Can also be seen with our own ruleset - change the catch converted back to use `IHasEndTime` and run `CatchBeatmapConversionTest`.

Would you be against making `IHasDuration : IHasEndTime` instead and using interface implementations to fix the setters?

```cs
/// <summary>
/// A HitObject that ends at a different time than its start time.
/// </summary>
#pragma warning disable 618
public interface IHasDuration : IHasEndTime
#pragma warning restore 618
{
    double IHasEndTime.EndTime
    {
        get => EndTime;
        set => Duration = (Duration - EndTime) + value;
    }

    double IHasEndTime.Duration => Duration;

    /// <summary>
    /// The time at which the HitObject ends.
    /// </summary>
    new double EndTime { get; }

    /// <summary>
    /// The duration of the HitObject.
    /// </summary>
    [JsonIgnore]
    new double Duration { get; set; }
}
```